### ISSUE TITLE
dotnet should support Debian distros

### ISSUE THREAD

**Issue by MichaelSimons** [dotnet/cli#212]
# Repro Steps
1. Create the following Dockerfile.

FROM buildpack-deps:jessie-scm

RUN echo "deb [arch=amd64] http://apt-mo.trafficmanager.net/repos/dotnet/ trusty main" > /etc/apt/sources.list.d/dotnetdev.list \
    && apt-key adv --keyserver apt-mo.trafficmanager.net --recv-keys 417A0893 \
    && echo "deb [arch=amd64] http://llvm.org/apt/trusty/ llvm-toolchain-trusty-3.6 main" > /etc/apt/sources.list.d/llvm.list \
    && wget -O - http://llvm.org/apt/llvm-snapshot.gpg.key|apt-key add - \
    && apt-get update \
    && apt-get install -y dotnet

RUN git clone https://<GitUsername>:<GitToken>@github.com/dotnet/cli.git cli

WORKDIR /cli/test/TestLibrary

RUN dotnet restore

WORKDIR /cli/test/TestApp

RUN dotnet restore

2. Replace <GitUsername> and <GitToken> with your credentials and build the docker image (using a Ubuntu 14.04-x64 machine) -  docker build -t restoreBug .
3. Run the built image - docker run -it restoreBug
4. Inspect the project.lock.json file that dotnet restore created.  Notice what runtimes it pulled packages for.
5.  Notice the error when you run dotnet run

# Actual Results
Notice that the two runtimes it pulled packages for are 8.debian-x64 and 8.debian-x86.   When you run step 6 you get the following error "Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'System.Console, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified."  To correct this behavior you can specify ubuntu.14.04-x64 as the runtime for commands like restore and publish.  Not sure of a workaround for run though.

# Expected Results
Previously whenever I ran these steps the dotnet was defaulting the runtime to as Ubuntu.14.04-x64.  It was not necessary to specify the --runtime parameter.

## Note
At the time this bug was logged, the version of the dotnet debian package used is 0.0.1-alpha-20151115-171248-1

**Comment by brthor** [comment]
Is this a debian docker image?

If so, it seems like 8.debian-x64 as the RID might be the correct behavior and we need to account for it in other places.

**Comment by Petermarcu** [comment]
Not much testing is happening on Debian right now. Most is on Ubuntu 14.04. Should we use that as the base image instead for now?

**Comment by brthor** [comment]
@Petermarcu This seems like the best solution for now, but we will need to think about this eventually. 

The Ubuntu base docker image is substantially larger than the debian image, so I think we'll want to go back at some point.

**Comment by piotrpMSFT** [comment]
Interesting. We should keep our images to Ubuntu for now. I want to get CI running on other platforms in the very near future but until then we can't guarantee anything about customer experience on non-Ubuntu.

**Comment by Petermarcu** [comment]
@brthor, absolutely. I'm just aiming to unblock right now. We will continue to add distros and deal with the differences throughout the stack.

**Comment by MichaelSimons** [comment]
Are there any thoughts on whether or not it would be appropriate to modify the RID detection code to default Debian distros to Ubuntu.  This defaulting seems backwards as it seems more appropriate to default Ubuntu to Debian.  If it is appropriate to change the RID detection logic, what should the version mapping logic be?  I assume the correct long term solution would to support Debian as an officially supported distro.

This issue is a significant issue for our Docker story.  In Docker land, Ubuntu is something mostly beginners use and not a common practice in production images. You can see almost all official Docker images (https://github.com/docker-library/) and our microsoft/aspnet image are based on Debian. Also read: http://mjg59.dreamwidth.org/36312.html

**Comment by glennc** [comment]
For the ASP.NET Docker image we set an environment variable to set the RID to ubuntu, kind of short circuiting RID detection to work on the debian base image. But I don't know if that variable made its way to dotnet yet. @anurse might.

I am comfortable doing that because we have made sure all the appropriate dependencies work and a bunch of people have been using it for a while. It is, however, theoretically possible that some future package would have an Ubuntu dependency that just doesn't work. Though I don't think that is likely.

**Comment by anurse** [comment]
This isn't actually Docker related at all, this is about running the CoreCLR on Debian (it just happens that most Docker images use Debian ;)). At the moment, CoreCLR only supports Ubuntu 14.04, not Debian. We can work around that in places, but what we really need is for CoreCLR to support Debian Jessie (or similar).

**Comment by adrian17** [comment]
I can still reproduce it on both Jessie and Stretch (future release) on a VM. May be connected to https://github.com/dotnet/cli/issues/291 ?

**Comment by MichaelSimons** [comment]
@anurse, yes this is not specific to Docker.  I also agree with you that we need CoreCLR to support Debian distros.  What is your opinion about providing a workaround in dotnet for the time being until Debian is officially supported?  We could either map Debian to Ubuntu (e.g. in Microsoft.Dnx.Tooling.RestoreRuntimeEnvironmentExtensions.GetDefaultRestoreRuntimes) or we could provide a hook such as an environment variable as @glennc mentions.  The DNX_RUNTIME_ID variable that works for ASP.NET/mono is not supported by dotnet but could easily be checked in places like Microsoft.Dnx.Tooling.RestoreCommand.GetRestoreRuntimes.  Adding a hook such as an environment variable might be more robust but it would require users to know about it and change the canonical usage pattern on Debian distros at least until it is officially supported.

**Comment by anurse** [comment]
I'd totally be in favor of a `DOTNET_RUNTIME_ID` environment variable fallback, and we can implement that sooner rather than later to get moving faster.

The most accurate solution would probably be to change the packages in dotnet/corefx to use an RID derived from Debian as much as possible (the big question is which, if any, version of `libicu` is provided by Debian; that's not the only one, but it is a major one). Once we have RIDs defining Debian, we could make this work in the "accurate" way /cc @ellismg @ericstj

I'd like to avoid mapping Debian to Ubuntu if possible, since the mapping should really go the other way around.

**Comment by MichaelSimons** [comment]
@anurse, I agree the mapping of Debian to Ubuntu would seem backwards.  If I understand the implementation correctly, the dotnet restore logic is contained within the shared DNX code.  If this is correct, it feels wrong to use an environment variable named 'DOTNET_RUNTIME_ID'.  Isn't this being refactored into CLI though? I prototyped a fix here (https://github.com/MichaelSimons/dnx/commit/d93c996c4a4f4fac55cf545ff57fe17a13055bbd).  Aside from the environment variable name question, is this inline with what you were thinking?

**Comment by anurse** [comment]
We do already have it in DNX: https://github.com/aspnet/dnx/blob/dev/src/Microsoft.Dnx.Runtime.Internals/RuntimeEnvironmentExtensions.cs#L15

The problem is that the parts of the code that read the lock file (in `Microsoft.DotNet.ProjectModel`) don't have a similar variable yet.

Eventually that code will be moving to `nuget` and `dotnet` though, so we need to think about where it should go then.

**Comment by spencerwi** [comment]
Another issue I'm hitting: `libicu52` is currently required by the the available deb packages, but in Ubuntu 15.10 (and presumably onwards) that package is superseded by the newer `libicu55`.

**Comment by piotrpMSFT** [comment]
@brthor take note of the libicu dependency

**Comment by brthor** [comment]
After reading this thread https://github.com/dotnet/cli/issues/674 I'm not sure coreclr/corefx will work with libicu55.

@ellismg You mentioned that System.Globalization requires specific versions of libicu. Would making this work be as simple as changing the dependency in the debian package or more involved on the framework side?

**Comment by ellismg** [comment]
@brthor You need to build System.Globalization from source against whatever ICU makes sense for the distro you are packaging for.  It will compile cleanly against any recent ICU.  What you can't do is take the binary version we built in CI for CentOS or Ubuntu 14.04 and then throw it on a different distro, since they might not have the same version of ICU.

So it's more than updating a dependency in a package, but you shouldn't need any code changes in the product, you just need to do a source build of the native components (which I would assume you are already doing?)

**Comment by brthor** [comment]
We do not currently orchestrate a build of the corefx native components. We consume the components as packaged in NuGet.

I'm not sure how feasible building these from source in the CLI repo is. It would increase our build times even further and would introduce complexity around ensuring the packages we build against and the native components we build ourselves are built from the same commit.

It seems like supporting libicu55 would be a prerequisite for corefx to support Ubuntu 15.10 anyways, so I assume at some point corefx would start producing packages with native components which are linked appropriately?

**Comment by ellismg** [comment]
> It seems like supporting libicu55 would be a prerequisite for corefx to support Ubuntu 15.10 anyways, so I assume at some point corefx would start producing packages with native components which are linked appropriately?

Yes.  If/when we support that RID, CoreFX would handle the packaging and building of the native shims.  If your build process just consumes packages based on RIDs for the target, you'd also be OK.

**Comment by brthor** [comment]
That we do. At that point, the only change required to produce a Ubuntu 15.10 compatible debian package would be creating an appropriate build definition (including ensuring we push to the right feed).

The updated libicu would be automatically picked up by the debian helper scripts used by the package tool.

**Comment by jeremymeng** [comment]
Any updates/workaround available now? I am getting errors when `dotnet restore` on Debian Jessie.
```
error: System.Diagnostics.Process 4.1.0-beta-23516 provides a compile-time reference assembly for System.Diagnostics.Process on DNXCore,Version=v5.0, but there is no run-time assembly compatible with debian.8-x64.
error: System.Globalization.Extensions 4.0.1-rc2-23805 provides a compile-time reference assembly for System.Globalization.Extensions on DNXCore,Version=v5.0, but there is no run-time assembly compatible with debian.8-x64.
```
`dotnet build` worked, but running the built executable throws exception
```
Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.

Aborted
```

**Comment by brthor** [comment]
@jeremymeng Try referencing the latest version of `NETStandard.Library` in your project.json

**Comment by jeremymeng** [comment]
@brthor I am still getting the same error after changing the version to `"NETStandard.Library": "1.0.0-rc3-*",`  I didn't mention that I have dependencies on ASP.NET RC2 bits.   They might depend on lower versions of coreclr?  I got many similar errors too, even for `rc3-23808` which seems to be the latest version.
```
error: System.IO 4.1.0-rc3-23808 provides a compile-time reference assembly for System.IO on DNXCore,Version=v5.0, but there is no run-time assembly compatible with debian.8-x64.
error: System.Linq.Expressions 4.0.11-rc3-23808 provides a compile-time reference assembly for System.Linq.Expressions on DNXCore,Version=v5.0, but there is no run-time assembly compatible with debian.8-x64.
```
if I use `--runtime ubuntu.14.04-x64`, I only got one error:
```
error: System.Linq.Expressions 4.0.11-rc3-23808 provides a compile-time reference assembly for System.Linq.Expressions on DNXCore,Version=v5.0, but there is no run-time assembly compatible with ubuntu.14.04-x64.
error: Some packages are not compatible with DNXCore,Version=v5.0 (ubuntu.14.04-x64).
```

**Comment by ellismg** [comment]
The debian issues are probably due to the fact that cli is going to compute "debian.8" as the RID for Debian Jessie.  Our [runtime.json](https://github.com/dotnet/corefx/blob/master/pkg/Microsoft.NETCore.Platforms/runtime.json#L224) graph does not list that RID (instead we have `debian` and `debian.8.2`).  `/etc/os-release` for Jessie lists ID as "8" so we should update our RID graph and packages to use that RID instead.

**Comment by piotrpMSFT** [comment]
@ellismg is there a related CoreFx issue?

**Comment by MichaelSimons** [comment]
@ellismg do you have an update on the CoreFX side?

**Comment by ellismg** [comment]
Opened dotnet/corefx#6170.  Should be an easy fix, the problem is a bunch of the packaging work is being transitioned from TFS to GH, so I don't know exactly where to make the fix this moment.  Chris will take a look.

**Comment by MichaelSimons** [comment]
@piotrpMSFT, What is the remaining work on supporting Debian?  It seems all corefx and rid issues have been resolved.  What installer/packaging changes are necessary?

**Comment by Petermarcu** [comment]
@MichaelSimons , you are trying to address this right now. Should we move this into rc2 or close it as a duplicate of whichever issue you are using to track debian package support?

**Comment by MichaelSimons** [comment]
@Petermarcu I moved it to rc2.

**Comment by piotrpMSFT** [comment]
Fixed by #2472

**Comment by piotrpMSFT** [comment]
@MichaelSimons please close when verified.
